cy.selectFile docs (#4241)
cypress-io/cypress-documentation: content/api/commands/attachfile.md

> ---

Reviewer: Be sure to add the new page to https://github.com/cypress-io/cypress-documentation/blob/master/content/_data/sidebar.json so it gets a spot in the list & API TOC.

Reviewer 2: You'll also need to add the new page to `content/_data/en.json`

Does `attachFile` also follow Actionability logic? If so we'll want to add it as a file that follows the logic to the `content/guides/core-concepts/interacting-with-elements.md` page.

Author: Added in all three places.

title: attachFile
---

> Attaches files to HTML5 input element, or simulated dragging files onto the browser.

Reviewer 2: ```suggestion
Attaches a file or files to an HTML5 input element or simulates dragging a file or files into the browser.
```

## Syntax

> ```javascript

Reviewer: This could be confusing since the method must be chained. I thinks it's ok to just indicate each as `.attachFile(foo)`, like we do in the `.within()` page.

Author: Good call, done.

cy.attachFile(file)
cy.attachFile(file, options)
cy.attachFile([file], options)
cy.attachFile([file], options)
```

### Usage

**<Icon name="check-circle" color="green"></Icon> Correct Usage**

```javascript
  cy.get('input[type=file]').attachFile('file.json')
  cy.get('input[type=file]').attachFile(['file.json', 'file2.json'])
  cy.get('input[type=file]').attachFile({
    contents: Buffer.from('foo'),
    fileName: 'file.json',
    lastModified: Date.now(),
  })
```

**<Icon name="exclamation-triangle" color="red"></Icon> Incorrect Usage**

```javascript
cy.attachFile('file.json') // Errors, cannot be chained off 'cy'

> cy.get('input[type=file]').attachFile('file contents') // Will attempt to find a file called 'file contents' on disk, probably not what you intended

Reviewer 2: ```suggestion

// Will attempt to find a file called 'file contents' on disk, probably not what you intended
cy.get('input[type=file]').attachFile('file contents') 
```

```

### Arguments

**<Icon name="angle-right"></Icon> file** **_(String, Array, Object or Buffer)_**

> Either a single file, or an array of them. A file can be:
> - A path to a file within the project root (the directory that contains the default `cypress.json` configuration file). Eg: `'path/to/file.json'`
> - `@alias` - An alias of any type, previously stored by `.as()`. Eg: `'@alias'`
> - A `Buffer()` containing binary data, such as that returned by `cy.readFile('file.json', { encoding: null })`. Eg: `Buffer.from('foo')`
> - An object with a non-null `contents` property, specifying details about the file. Eg: `{contents: '@alias', fileName: 'file.json'}`

Reviewer 2: Avoid pronouns, instead be more specific to avoid what 'them' is referring to. 

```suggestion
Either a single file, or an array of files. A file can be:

- A path to a file within the project root (the directory that contains the default Cypress configuration file). Eg: `'path/to/file.json'`
- `@alias` - An alias of any type, previously stored using `.as()`. Eg: `'@alias'`
- A `Buffer()` containing binary data, such as that returned by `cy.readFile('file.json', { encoding: null })`. Eg: `Buffer.from('foo')`
- An object with a non-null `contents` property, specifying details about the file. Eg: `{contents: '@alias', fileName: 'file.json'}`
```

If an object is provided, it can have the following properties.

| Option         | Description                                                                                                                                                                                                              |
| -------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| `contents`     | The contents of the file. This can be a string shorthand as described above, a `Buffer()` containing binary data or a non-Buffer object, which will be converted into a string with `JSON.stringify()` and utf8 encoded. |
| `fileName`     | The name of the file. If `contents` is a path on disk, this defaults to the actual filename. In any other case, this defaults to an empty string.                                                                        |

> | `lastModified` | The file's last modified timestamp. This defaults to `Date.now()`.                                                                                                                                                       |

Reviewer 2: Might be good to clarify the time format expected (I guess saying Date.now() kind of does that). I can just see people passing in different datetime formats or something and not understanding why it's not working.

Author: Done. Since I keep having to look up exactly how to get the timestamp from a js Date, I'm just adding a MDN link straight to the docs.

| `action`            | `'input'`                                                             | Switches attachFile between modes. Valid values are `input` and `drag-n-drop`. See [Actions](#Actions) below for more details.                                      |
| `animationDistanceThreshold` | [`animationDistanceThreshold`](/guides/references/configuration#Actionability) | The distance in pixels an element must exceed over time to be [considered animating](/guides/core-concepts/interacting-with-elements#Animations). |
| `force`             | `false`                                                               | Forces the action, disables [waiting for actionability](#Assertions)                                                                                                |
| `log`               | `true`                                                                | Displays the command in the [Command log](/guides/core-concepts/test-runner#Command-Log)                                                                            |

> | `timeout`           | [`defaultCommandTimeout`](/guides/references/configuration#Timeouts)  | Time to wait for `.find()` to resolve before [timing out](#Timeouts)                                                                                                |

Reviewer 2: typo

```suggestion
| `timeout`           | [`defaultCommandTimeout`](/guides/references/configuration#Timeouts)  | Time to wait for `.attachFile()` to resolve before [timing out](#Timeouts)                                                                                                |
```

| `waitForAnimations` | [`waitForAnimations`](/guides/references/configuration#Actionability) | Whether to wait for elements to [finish animating](/guides/core-concepts/interacting-with-elements#Animations) before executing the command.                        |

### Yields [<Icon name="question-circle"/>](/guides/core-concepts/introduction-to-cypress#Subject-Management)

<List><li>`.attachFile()` yields the same subject it was given from the previous command.</li></List>

### Action

> Depending on the action set in the `options` argument, `cy.attachFile()` can simulate two different user behaviors:

Reviewer: ```suggestion
Depending on the action set in the `options` argument, `.attachFile()` can simulate two different user behaviors:
```

> - `input` (default)
> By default, `attachFile()` runs in 'input' mode, mimicking a user a selecting one or more files on an HTML5 input element. In this mode, the subject [<Icon name="question-circle"/>](/guides/core-concepts/introduction-to-cypress#Subject-Management) must be a single `input` element with `type="file"`, or a `label` element connected to one (either with its `for` attribute or by containing the input).
> - `drag-n-drop`
> Setting the action to `drag-n-drop` changes the behavior of the command to instead mimic a user dragging files from the operating system into the browser, and dropping them over the selected subject. In this mode, the subject can be any DOM element, or the `document` as a whole.

Reviewer 2: ```suggestion
- `input` (default)
By default, `attachFile()` runs in 'input' mode, mimicking a user selecting one or more files on an HTML5 input element. In this mode, the subject [<Icon name="question-circle"/>](/guides/core-concepts/introduction-to-cypress#Subject-Management) must be a single `input` element with `type="file"`, or a `label` element connected to an input (either with its `for` attribute or by containing the input).
- `drag-n-drop`
Setting the action to `drag-n-drop` changes the behavior of the command to instead mimic a user dragging files from the operating system into the browser, and dropping them over the selected subject. In this mode, the subject can be any DOM element or the `document` as a whole.
```

## Examples

### From a file on disk

```javascript
cy.get('input[type=file]').attachFile('file.json')
cy.get('input[type=file]').attachFile('file.png')
```

> If given a path, `cy.readFile()` will attach the file exactly as it exists on disk. This is the preferred way to work with files on disk, avoiding many encoding-related pitfalls.

Reviewer: ```suggestion
If given a path, `.attachFile()` will attach the file exactly as it exists on disk. This is the preferred way to work with files on disk, avoiding many encoding-related pitfalls.
```

> If given a path, `cy.readFile()` will attach the file exactly as it exists on disk. This is the preferred way to work with files on disk, avoiding many encoding-related pitfalls.

Reviewer 2: typo

```suggestion
If given a path, `.attachFile()` will attach the file exactly as it exists on disk. This is the preferred way to work with files on disk, avoiding many encoding-related pitfalls.
```

### On a hidden input

```javascript
cy.get('input[type=file]').attachFile('file.json', { force: true })
```

> In many cases in modern applications, the underlaying file input is hidden from view, and activated by a user clicking on a button. In these cases, you will need to tell Cypress to ignore its actionability checks and attach the file even though a user would not be able to directly activate the file input.

Reviewer 2: typo

```suggestion
In many cases in modern applications, the underlying file input is hidden from view, and activated by a user clicking on a button. In these cases, you will need to tell Cypress to ignore its actionability checks and attach the file even though a user would not be able to directly activate the file input.
```

### From a fixture
```javascript
cy.fixture('file.json', { encoding: null }).as('myFixture')
cy.get('input[type=file]').attachFile('@myFixture')
```

> Note the use of `null` encoding. By default, `cy.fixture()` `cy.readFile()` attempt to interpret files read from disk, which would result in a json file being decoded and re-encoded as a utf-8 string - the contents would be preserved, but formatting would not be, and the encoding might change. See [`cy.fixture`](/api/commands/fixture) or [`cy.readFile`](/api/commands/readfile) for more details on file encoding.

Reviewer 2: ```suggestion
Note the use of `null` encoding. By default, `cy.fixture()` and `cy.readFile()` attempt to interpret files read from disk, which would result in a JSON file being decoded and re-encoded as a utf-8 string - the contents would be preserved, but formatting would not be and the encoding might change. See [`cy.fixture`](/api/commands/fixture) or [`cy.readFile`](/api/commands/readfile) for more details on file encoding.
```

### From an API response

```javascript
cy.request('http://localhost:8888/users/827').get('body').as('responseBody')
cy.get('input[type=file]').attachFile('@responseBody')
```

### Processing data inside the test

```javascript
cy.readFile('users.json').then((users => {
  users[0].username = 'JohnCena'
})).as('myFile')

cy.get('input[type=file]').attachFile('@myFile')
```

### Attaching multiple files

```javascript
cy.get('input[type=file]').attachFile(['file1.json', 'file2.json', 'file3.json'])
```

This will fail unless the file input has the `multiple` property.

### Attaching a file with custom filename and lastModified

```javascript
cy.get('input[type=file]').attachFile({
  contents: 'path/to/file.json',
  fileName: 'users.json',
  lastModified: new Date('Feb 18 1989').valueOf()
})
```

### Dropping a file on the document

```javascript
cy.document().attachFile('file.json', { action: 'drag-n-drop' })
```

## Notes

### Existence

#### Default file existence assertion

> When passed a file path, `cy.attachFile()` asserts that the file exists and will fail if it does not exist. It will retry reading the file if it does not initially exist until the file exists or the command times out.

Reviewer: ```suggestion
When passed a file path, `.attachFile()` asserts that the file exists and will fail if it does not exist. It will retry reading the file if it does not initially exist until the file exists or the command times out.
```

```javascript
// will fail after the defaultCommandTimeout is reached
cy.get('input[type=file]').attachFile('does-not-exist.yaml')
```

### Actionability

#### The element must first reach actionability

`.attachFile()` is an "action command" that follows all the rules
[defined here](/guides/core-concepts/interacting-with-elements).

## Rules

### Requirements [<Icon name="question-circle"/>](/guides/core-concepts/introduction-to-cypress#Chains-of-Commands)

<List><li>`.attachFile()` requires being chained off a command that yields DOM

> element(s). With the `input` action (default), it further requires a single `input` element with `type="file"`, or a `label` element attached to one.</li><li>If given a path, `cy.attachFile()` requires the file must exist.</li><li>If given an alias, `cy.attachFile()` requires that the subject of the alias must not be `null` or `undefined`</li></List>

Reviewer: ```suggestion
element(s). With the `input` action (default), it further requires a single `input` element with `type="file"`, or a `label` element attached to one.</li><li>If given a path, `.attachFile()` requires the file must exist.</li><li>If given an alias, `.attachFile()` requires that the subject of the alias must not be `null` or `undefined`</li></List>
```
```suggestion
element(s). With the `input` action (default), it further requires a single `input` element with `type="file"`, or a `label` element attached to one.</li><li>If given a path, `.attachFile()` requires the file must exist.</li><li>If given an alias, `.attachFile()` requires that the subject of the alias must not be `null` or `undefined`</li></List>
```

### Assertions [<Icon name="question-circle"/>](/guides/core-concepts/introduction-to-cypress#Assertions)

<List><li>`.attachFile()` will automatically wait for the element to reach an
[actionable state](/guides/core-concepts/interacting-with-elements)</li></List>

### Timeouts [<Icon name="question-circle"/>](/guides/core-concepts/introduction-to-cypress#Timeouts)

<List><li>`.attachFile()` can time out waiting for the element to reach an
[actionable state](/guides/core-concepts/interacting-with-elements).</li></List>

## Command Log

**_Attach file to input_**

```javascript
cy.get('.file-input').attachFile(Buffer.from('Hello world'))
```

The commands above will display in the Command Log as:

<DocsImage src="/img/api/attachfile/attach-file-during-test.png" alt="Command log for attachFile" ></DocsImage>

> When clicking on `click` within the command log, the console outputs the

Reviewer 2: typo

```suggestion
When clicking on `attachFile` within the command log, the console outputs the
```

following:

> <DocsImage src="/img/api/click/attach-file-in-console.png" alt="console.log for attachFile" ></DocsImage>

Reviewer: Did you want this image in the click folder? Will need either an updated path or to pop the png in `/attachfile`

Author: Just a copy paste error, fixed.

## History

| Version                                     | Changes                                                  |
| ------------------------------------------- | -------------------------------------------------------- |

> | [10.0.0](/guides/references/changelog#10.0.0) | `cy.attachFile()` command added                        |

Reviewer: ```suggestion
| [10.0.0](/guides/references/changelog#10.0.0) | `.attachFile()` command added                        |
```

> | [10.0.0](/guides/references/changelog#10.0.0) | `cy.attachFile()` command added                        |

Reviewer 2: Does this have to go into 10.0.0? I feel like this can go in an earlier release. Usually we wait to add the History table until we have the release branch.

```suggestion
| [10.0.0](/guides/references/changelog#10.0.0) | `cy.attachFile()` command added                        |
```

Author: So the reason I was thinking 10.0 is because of the popularity of the existing community solution, and the fact that adding a command that shares a name will throw an error if the plugin is still installed. While we did discuss 'not every fix is breaking' in standup today, and I agree it can sometimes be murky, but I think this one should be in a major version.

cypress-upload-file has [762 thousand](https://www.npmjs.com/package/cypress-file-upload) (!!!) weekly downloads - that's more a quarter as many downloads as Cypress itself. When we add in our own version of this command, every single one of those user's tests will break and they'll need to follow a migration guide.

It may theoretically be a new feature (minor release), but in practice it'll require a large portion of our users to update their code.

Reviewer 3: If this is going into 10.0.0, it would be nice to see it based off of the `unification` branch. Otherwise, carry on!

Author: Switched to targeting the unification branch, thanks. I was looking for a 10.0-release branch matching the name of the cypress branch, or the 9.2-release branch naming we have in this repo and didn't see it.

## See also

- [Guide: Variables and Aliases](/guides/core-concepts/variables-and-aliases)
- [`.get()`](/api/commands/get)

> - [`.readFile()`](/api/commands/readfile)

Reviewer 2: cy.fixture

Author: Added.

| `log`               | `true`                                                                | Displays the command in the [Command log](/guides/core-concepts/test-runner#Command-Log)                                                                            |
| `timeout`           | [`defaultCommandTimeout`](/guides/references/configuration#Timeouts)  | Time to wait for `.attachFile()` to resolve before [timing out](#Timeouts)                                                                                                |
| `waitForAnimations` | [`waitForAnimations`](/guides/references/configuration#Actionability) | Whether to wait for elements to [finish animating](/guides/core-concepts/interacting-with-elements#Animations) before executing the command.                        |

### Yields [<Icon name="question-circle"/>](/guides/core-concepts/introduction-to-cypress#Subject-Management)

<List><li>`.attachFile()` yields the same subject it was given from the previous command.</li></List>

### Action

Depending on the action set in the `options` argument, `.attachFile()` can simulate two different user behaviors:

> - `input` (default)

Reviewer 2: The formatting of this looks strange when rendered. I think these should be H4s, or I think you need to indent the content within the li. See preview: https://deploy-preview-4241--cypress-docs.netlify.app/api/commands/attachfile#Arguments

> By default, `attachFile()` runs in 'input' mode, mimicking a user selecting one or more files on an HTML5 input element. In this mode, the subject [<Icon name="question-circle"/>](/guides/core-concepts/introduction-to-cypress#Subject-Management) must be a single `input` element with `type="file"`, or a `label` element connected to an input (either with its `for` attribute or by containing the input).

Reviewer 2: This icon is not standard to use inline with sentences, we just link the word. 

```suggestion
By default, `attachFile()` runs in 'input' mode, mimicking a user selecting one or more files on an HTML5 input element. In this mode, the [subject](/guides/core-concepts/introduction-to-cypress#Subject-Management) must be a single `input` element with `type="file"`, or a `label` element connected to an input (either with its `for` attribute or by containing the input).
```

Author: > A few other things I noticed while reading through the updated preview.
> 
> Some of the code blocks are showing scrollbars at the max size. Can you update the code blocks so you don't have to scroll to read? See preview: [deploy-preview-4241--cypress-docs.netlify.app/api/commands/attachfile#Arguments](https://deploy-preview-4241--cypress-docs.netlify.app/api/commands/attachfile#Arguments)

Went through and fixed all those.

Author: Ah, whoop, last one wasn't supposed to be a child comment. Also fixed this.

### On a hidden input

```javascript
cy.get('input[type=file]').attachFile('file.json', { force: true })
```

In many cases in modern applications, the underlying file input is hidden from view, and activated by a user clicking on a button. In these cases, you will need to tell Cypress to ignore its actionability checks and attach the file even though a user would not be able to directly activate the file input.

### From a fixture
```javascript
cy.fixture('file.json', { encoding: null }).as('myFixture')
cy.get('input[type=file]').attachFile('@myFixture')
```

Note the use of `null` encoding. By default, `cy.fixture()` and `cy.readFile()` attempt to interpret files read from disk, which would result in a JSON file being decoded and re-encoded as a utf-8 string - the contents would be preserved, but formatting would not be and the encoding might change. See [`cy.fixture`](/api/commands/fixture) or [`cy.readFile`](/api/commands/readfile) for more details on file encoding.

### From an API response

```javascript

> cy.request('http://localhost:8888/users/827').get('body').as('responseBody')

Reviewer 2: This isn't correct. `cy.get()` cannot be chained off of another command. It's a parent command.

Author: You're right, the example is incorrect. Oddly though, it doesn't throw an error:
```
cy.request('http://example.com').get('body').then(a => { console.log(a) })
// a is now the <body> element of the AUT
```
Maybe something to consider fixing with `.get()` at some point, making it stricter.

Will fix the example.

Reviewer 2: Yah, we never got around to enforced chaining. That would be a major breaking change.

---
title: attachFile
---

Attaches a file or files to an HTML5 input element or simulates dragging a file or files into the browser.

## Syntax

```javascript
.attachFile(file)
.attachFile(file, options)
.attachFile([file], options)
.attachFile([file], options)
```

### Usage

**<Icon name="check-circle" color="green"></Icon> Correct Usage**

> ```javascript

Reviewer: We can get a little more of this section in view by removing the extra couple of spaces in front of this fenced in block.

```suggestion
```javascript
cy.get('input[type=file]').attachFile('file.json')
cy.get('input[type=file]').attachFile(['file.json', 'file2.json'])
cy.get('input[type=file]').attachFile({
  contents: Buffer.from('foo'),
  fileName: 'file.json',
  lastModified: Date.now(),
})
```

Reviewer: This one may have gotten wonky via GH suggestions with the multiple sets of ```. The idea is just to close the extra white space.

Author: Having the spaces looks nicer in the markdown, but markdown isn't the actual target format. :)

Done.

**<Icon name="angle-right"></Icon> options** **_(Object)_**

Pass in an options object to change the default behavior of `.attachFile()`.

| Option              | Default                                                               | Description                                                                                                                                                         |
| ------------------- | --------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `action`            | `'input'`                                                             | Switches attachFile between modes. Valid values are `input` and `drag-n-drop`. See [Actions](#Actions) below for more details.                                      |
| `animationDistanceThreshold` | [`animationDistanceThreshold`](/guides/references/configuration#Actionability) | The distance in pixels an element must exceed over time to be [considered animating](/guides/core-concepts/interacting-with-elements#Animations). |

> | `force`             | `false`                                                               | Forces the action, disables [waiting for actionability](#Assertions)                                                                                                |

Reviewer: ```suggestion
| `force`             | `false`                                                               | Forces the action, disables [waiting for actionability](#Assertions).                                                                                                |
```

> | `log`               | `true`                                                                | Displays the command in the [Command log](/guides/core-concepts/test-runner#Command-Log)                                                                            |

Reviewer: ```suggestion
| `log`               | `true`                                                                | Displays the command in the [Command log](/guides/core-concepts/test-runner#Command-Log).                                                                            |
```

> | `timeout`           | [`defaultCommandTimeout`](/guides/references/configuration#Timeouts)  | Time to wait for `.attachFile()` to resolve before [timing out](#Timeouts)                                                                                                |

Reviewer: ```suggestion
| `timeout`           | [`defaultCommandTimeout`](/guides/references/configuration#Timeouts)  | Time to wait for `.attachFile()` to resolve before [timing out](#Timeouts).                                                                                                |
```

### From a fixture
```javascript
cy.fixture('file.json', { encoding: null }).as('myFixture')
cy.get('input[type=file]').attachFile('@myFixture')
```

Note the use of `null` encoding. By default, `cy.fixture()` and `cy.readFile()` attempt to interpret files read from disk, which would result in a JSON file being decoded and re-encoded as a utf-8 string - the contents would be preserved, but formatting would not be and the encoding might change. See [`cy.fixture`](/api/commands/fixture) or [`cy.readFile`](/api/commands/readfile) for more details on file encoding.

### From an API response

```javascript
cy.request('http://localhost:8888/users/827')
  .its('body')
  .as('responseBody')

cy.get('input[type=file]').attachFile('@responseBody')
```

### Processing data inside the test

```javascript
cy.readFile('users.json').then((users => {
  users[0].username = 'JohnCena'
})).as('myFile')

cy.get('input[type=file]').attachFile('@myFile')
```

### Attaching multiple files

```javascript
cy.get('input[type=file]').attachFile([
  'file1.json',
  'file2.json',
  'file3.json',
])
```

> This will fail unless the file input has the `multiple` property.

Reviewer: This is probably worth throwing in a lil alert to make sure it doesn't get lost within in the content.

```suggestion
<Alert type="warning">

This will fail unless the file input has the `multiple` property.

</Alert>
```

### Attaching a file with custom filename and lastModified

```javascript
cy.get('input[type=file]').attachFile({
  contents: 'path/to/file.json',
  fileName: 'users.json',
  lastModified: new Date('Feb 18 1989').valueOf()
})
```

### Dropping a file on the document

```javascript
cy.document().attachFile('file.json', { action: 'drag-n-drop' })
```

## Notes

### Existence

#### Default file existence assertion

When passed a file path, `.attachFile()` asserts that the file exists and will fail if it does not exist. It will retry reading the file if it does not initially exist until the file exists or the command times out.

```javascript
// will fail after the defaultCommandTimeout is reached
cy.get('input[type=file]').attachFile('does-not-exist.yaml')
```

### Actionability

#### The element must first reach actionability

`.attachFile()` is an "action command" that follows all the rules
[defined here](/guides/core-concepts/interacting-with-elements).

## Rules

### Requirements [<Icon name="question-circle"/>](/guides/core-concepts/introduction-to-cypress#Chains-of-Commands)

<List><li>`.attachFile()` requires being chained off a command that yields DOM

> element(s). With the `input` action (default), it further requires a single `input` element with `type="file"`, or a `label` element attached to one.</li><li>If given a path, `.attachFile()` requires the file must exist.</li><li>If given an alias, `.attachFile()` requires that the subject of the alias must not be `null` or `undefined`</li></List>

Reviewer: ```suggestion
element(s). With the `input` action (default), it further requires a single `input` element with `type="file"`, or a `label` element attached to one.</li><li>If given a path, `.attachFile()` requires the file must exist.</li><li>If given an alias, `.attachFile()` requires that the subject of the alias must not be `null` or `undefined`.</li></List>
```

### Assertions [<Icon name="question-circle"/>](/guides/core-concepts/introduction-to-cypress#Assertions)

<List><li>`.attachFile()` will automatically wait for the element to reach an

> [actionable state](/guides/core-concepts/interacting-with-elements)</li></List>

Reviewer: ```suggestion
[actionable state](/guides/core-concepts/interacting-with-elements).</li></List>
```

Author: I sure do love my missing periods!

---
title: attachFile
---

Attaches a file or files to an HTML5 input element or simulates dragging a file or files into the browser.

## Syntax

```javascript
.attachFile(file)
.attachFile(file, options)
.attachFile([file], options)

> .attachFile([file], options)

Reviewer 3: I feel like the `.attachFile([file], options)` lines are unnecessary here

Author: There definitely shouldn't be two of them, I'll fix that.

Even though it's not strictly speaking a different signature (still 1-2 args), it seemed worth highlighting, right at the top, that you can pass in an array of files. Perhaps `.attachFile([file, file2, ...], options)` would be clearer in that intention? Or do you not think it's worth having additional lines here at all?

Reviewer 3: I think we could do

```js
.attachFile([file, file2, ...], options)
```

or 

```js
.attachFile(fileArray, options)
```

but in either case, we probably also want to show the non-`options` variant, so maybe something like this?

```js
.attachFile(file)
.attachFile(file, options)
.attachFile(fileArray)
.attachFile(fileArray, options)
```

Author: Yeah, that's why there was a duplicate line, accidentally had `options` in both. Pushed up changes.

| `log`               | `true`                                                                | Displays the command in the [Command log](/guides/core-concepts/test-runner#Command-Log).                                                                           |
| `timeout`           | [`defaultCommandTimeout`](/guides/references/configuration#Timeouts)  | Time to wait for `.attachFile()` to resolve before [timing out](#Timeouts).                                                                                               |
| `waitForAnimations` | [`waitForAnimations`](/guides/references/configuration#Actionability) | Whether to wait for elements to [finish animating](/guides/core-concepts/interacting-with-elements#Animations) before executing the command.                        |

### Yields [<Icon name="question-circle"/>](/guides/core-concepts/introduction-to-cypress#Subject-Management)

<List><li>`.attachFile()` yields the same subject it was given from the previous command.</li></List>

### Action

> Depending on the action set in the `options` argument, `.attachFile()` can simulate two different user behaviors:

Reviewer 3: ```suggestion
Depending on the `action` set in the `options` argument, `.attachFile()` can simulate two different user behaviors:
```

#### `input` (default)

By default, `attachFile()` runs in 'input' mode, mimicking a user selecting one or more files on an HTML5 input element. In this mode, the [subject](/guides/core-concepts/introduction-to-cypress#Subject-Management) must be a single `input` element with `type="file"`, or a `label` element connected to an input (either with its `for` attribute or by containing the input).

#### `drag-n-drop`

Setting the action to `drag-n-drop` changes the behavior of the command to instead mimic a user dragging files from the operating system into the browser, and dropping them over the selected subject. In this mode, the subject can be any DOM element or the `document` as a whole.

## Examples

### From a file on disk

```javascript

> cy.get('input[type=file]').attachFile('file.json')

Reviewer 3: It might be helpful to reinforce that the file can have a path

```suggestion
cy.get('input[type=file]').attachFile('path/to/file.json')
```

Author: Done.

cy.get('input[type=file]').attachFile('file.png')
```

> If given a path, `.attachFile()` will attach the file exactly as it exists on disk. This is the preferred way to work with files on disk, avoiding many encoding-related pitfalls.

Reviewer 3: Also reinforcing that files are loaded relative to the project root

```suggestion
If given a path, `.attachFile()` will attach the file, relative to the project root, exactly as it exists on disk. This is the preferred way to work with files on disk, avoiding many encoding-related pitfalls.
```

### From an API response

```javascript
cy.request('http://localhost:8888/users/827')
  .its('body')
  .as('responseBody')

cy.get('input[type=file]').attachFile('@responseBody')
```

### Processing data inside the test

```javascript
cy.readFile('users.json').then((users => {
  users[0].username = 'JohnCena'
})).as('myFile')

cy.get('input[type=file]').attachFile('@myFile')
```

### Attaching multiple files

```javascript
cy.get('input[type=file]').attachFile([
  'file1.json',
  'file2.json',
  'file3.json',
])
```

<Alert type="warning">

This will fail unless the file input has the `multiple` property.

</Alert>

### Attaching a file with custom filename and lastModified

```javascript
cy.get('input[type=file]').attachFile({
  contents: 'path/to/file.json',
  fileName: 'users.json',
  lastModified: new Date('Feb 18 1989').valueOf()
})
```

### Dropping a file on the document

```javascript
cy.document().attachFile('file.json', { action: 'drag-n-drop' })
```

## Notes

### Existence

#### Default file existence assertion

> When passed a file path, `.attachFile()` asserts that the file exists and will fail if it does not exist. It will retry reading the file if it does not initially exist until the file exists or the command times out.

Reviewer 3: Does it also do this with an object with a `contents` property with a file path string?

Author: Yep. Clarified slightly.

```javascript
// will fail after the defaultCommandTimeout is reached
cy.get('input[type=file]').attachFile('does-not-exist.yaml')
```

### Actionability

#### The element must first reach actionability

`.attachFile()` is an "action command" that follows all the rules

> [defined here](/guides/core-concepts/interacting-with-elements).

Reviewer 3: ```suggestion
of [Actionability](/guides/core-concepts/interacting-with-elements#Actionability).
```

Author: I copied this directly from another command, and it seems to be the common pattern in our docs.

![image](https://user-images.githubusercontent.com/3003404/146038924-2b69cf3b-a210-445b-a2aa-105b6ec5b90e.png)

Do you want to change this everywhere?

Reviewer 3: I think we should change it everywhere, but that can also be done in a separate, follow-up PR instead of further complicating this one. If we do that, just leave it as-is. Your call!

Author: It's a good change but a small one, so I went ahead an applied it to all the docs here.

## Rules

### Requirements [<Icon name="question-circle"/>](/guides/core-concepts/introduction-to-cypress#Chains-of-Commands)

> <List><li>`.attachFile()` requires being chained off a command that yields DOM

Reviewer 3: Can these be markdown lists?

Author: I am not exactly sure why they're not, but I was copying the pattern from other files in this directory - they universally use `<List>` in their Rules section. Without understanding Chesterton's fence, I didn't want to change it. I'm wondering if maybe some of our custom scripts rely on it?

![image](https://user-images.githubusercontent.com/3003404/146043293-ac22fc7d-7348-4df7-abce-fe33296defa2.png)

Reviewer 3: I'm not entirely sure what's up with `<List>`, but the `List` component is just this:

```vue
<template>
  <ul>
    <slot />
  </ul>
</template>
```

I loaded all the commit diffs containing `<List` into vscode with:

```sh
git log -G'<List' -p | code -
```

and then I searched for `(?<![</])\bList\b` to try to find any references to `List` that aren't preceded by `<` or `/` (to filter out `<List>` and `</List>` and found nothing of interest.

I think we're ok making it a markdown list.

### Timeouts [<Icon name="question-circle"/>](/guides/core-concepts/introduction-to-cypress#Timeouts)

<List><li>`.attachFile()` can time out waiting for the element to reach an
[actionable state](/guides/core-concepts/interacting-with-elements).</li><li>
`.attachFile()` can time out waiting for a file to exist on disk or for an
alias to resolve.</li></List>

## Command Log

**_Attach file to input_**

```javascript
cy.get('.file-input').attachFile(Buffer.from('Hello world'))
```

The commands above will display in the Command Log as:

<DocsImage src="/img/api/attachfile/attach-file-during-test.png" alt="Command log for attachFile" ></DocsImage>

When clicking on `attachFile` within the command log, the console outputs the
following:

<DocsImage src="/img/api/attachfile/attach-file-in-console.png" alt="console.log for attachFile" ></DocsImage>

## History

| Version                                       | Changes                                                  |
| --------------------------------------------- | -------------------------------------------------------- |
| [10.0.0](/guides/references/changelog#10.0.0) | `.attachFile()` command added                            |

> ### Community Recognition

Author: @jennifer-shehane , @cowboy , @jaffrepaul - Calling this section out as it's newly added since the last round of reviews. Just cleaning up some loose ends.

We don't really have a standard for this, but I borrowed the format from https://github.com/cypress-io/cypress/blob/develop/npm/cypress-schematic/README.md#community-recognition.

## History

| Version                                       | Changes                                                  |
| --------------------------------------------- | -------------------------------------------------------- |
| [10.0.0](/guides/references/changelog#10.0.0) | `.attachFile()` command added                            |

### Community Recognition

> The `.attachFile()` command draws heavy inspiration from the now-deprecated Cypress File Upload plugin. It was make possible by contributors to the [@abramenal/cypress-file-upload](https://github.com/abramenal/cypress-file-upload) repository.

Reviewer: Typo in there. _It was made possible..._
